core/shared/platform/include/platform_api_extension.h

```diff
@@ -1111,6 +1111,472 @@ os_socket_get_broadcast(bh_socket_t socket, bool *is_enabled);
 int
 os_dumps_proc_mem_info(char *out, unsigned int size);
 
+#if WASM_ENABLE_LIBC_WASI != 0 && WASM_ENABLE_UVWASI != 1
+
+#include "wasmtime_ssp.h"
```

Comment:
I wonder if adding this file doesn't introduce a logical circular dependency - we add dependency to a header that is part of libc-wasi implementation, but wasi-libc implementation itself depends on the platform code. I know the code compiles, just wanted to make sure we don't break anything on the conceptual level.

@wenyongh / @yamt any thoughts?

Reply:
> but wasi-libc implementation itself depends on the platform code

the wasi-libc _implementation_ (i.e. posix.c) does depend on platform code but `wasmtime_ssp.h` itself doesn't depend on any platform code - it's a header which defines the internal wasi-libc interface + set of wasi types. It doesn't depend on any internal headers (except wasm_export.h which itself is a standalone header by necessity) so it should be safe for anything else to depend on it without introducing a circular dependency.

Reply:
Yes, that's really an issue, at least it doesn't comply with the design rule. In our original design, the platform layer is the lowest layer, it shouldn't rely on any other layers (except it can call libc API). It seems that this file requires wasi types, such as __wasi_errno_t, __wasi_filestat_t, __wasi_iovec_t and so on. My understanding is that we should also move them into the platform layer, and don't include wasmtime_ssp.h.

And there are similar issues in posix_file.h, which includes bh_common.h and wasi_errno.h.

@loganek what's your opinion?

Reply:
> > but wasi-libc implementation itself depends on the platform code
> 
> the wasi-libc _implementation_ (i.e. posix.c) does depend on platform code but `wasmtime_ssp.h` itself doesn't depend on any platform code - it's a header which defines the internal wasi-libc interface + set of wasi types. It doesn't depend on any internal headers (except wasm_export.h which itself is a standalone header by necessity) so it should be safe for anything else to depend on it without introducing a circular dependency.

Got it, how about moving wasmtime_ssp.h to this folder, remove wasm_export.h in it, and create another file head file to declare these APIs, and then include it in platform_api_extension.h to avoid change a lot in the latter?

Reply:
Yes, that sounds good to me. 

Just want to to check I'm understanding correctly - we split `wasmtime_ssp.h` into 2 parts: one header with just the WASI types and another header with the wasmtime_ssp functions since the platform layer doesn't need to know about these functions.

Something like (apologies if this isn't super clear):

`core/shared/platform/include/platform_wasi_types.h` -> standalone header containing wasi types which platform implementations as well as wasi-libc implementation can depend on. Not sure about the naming of this header.

`core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/include/wasmtime_ssp.h` -> contains wasmtime_ssp API (e.g. `wasmtime_ssp_fd_renumber`, `wasmtime_ssp_fd_seek`) which only the wasi-libc implementation will depend on. (this header itself would depend on the `platform_wasi_types.h`.

The only reason I used wasi types in all the filesystem functions is because they're necessarily platform-agnostic and it feels like a lot of boilerplate to define `bh_` types which do the same thing, `bh_filestat_t`, `bh_filesize_t` etc. But that would be another option.

> And there are similar issues in posix_file.h, which includes bh_common.h and wasi_errno.h.

`bh_common.h` was necessary for `bh_memcpy_s` but I assume we can use `memcpy` instead - I don't know if there's any advantage/disadvantage. Not sure about the best place for `wasi_errno.h` yet.

Reply:
> Yes, that sounds good to me.
> 
> Just want to to check I'm understanding correctly - we split `wasmtime_ssp.h` into 2 parts: one header with just the WASI types and another header with the wasmtime_ssp functions since the platform layer doesn't need to know about these functions.
> 
> Something like (apologies if this isn't super clear):
> 
> `core/shared/platform/include/platform_wasi_types.h` -> standalone header containing wasi types which platform implementations as well as wasi-libc implementation can depend on. Not sure about the naming of this header.

How about changing to platform_wasi.h, moving the current wasi API declarations from platform_api_extension.h to it (so it includes wasi types and apis), and letting platform_api_extension.h include it?

> 
> `core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/include/wasmtime_ssp.h` -> contains wasmtime_ssp API (e.g. `wasmtime_ssp_fd_renumber`, `wasmtime_ssp_fd_seek`) which only the wasi-libc implementation will depend on. This header itself would depend on `platform_wasi_types.h`.
> 

It is ok to me.

> The only reason I used wasi types in all the filesystem functions is because they're necessarily platform-agnostic and it feels like a lot of boilerplate to define `bh_` types which do the same thing, `bh_filestat_t`, `bh_filesize_t` etc. But that would be another option.

Seems better to keep unchanged (use the current __wasi prefix)?

> 
> > And there are similar issues in posix_file.h, which includes bh_common.h and wasi_errno.h.
> 
> `bh_common.h` was necessary for `bh_memcpy_s` but I assume we can use `memcpy` instead - I don't know if there's any advantage/disadvantage. Not sure about the best place for `wasi_errno.h` yet.

It should be ok to use memcpy in platform as soon as we can ensure the security. bh_memcpy_s adds more checks but  it should be not mandatory in platform layer.

Reply:
Yes, I agree with the approach above in general. My main problem was that we've included wasmtime_ssp.h file that was part of the wasi-libc library (even though it was just a header file).

Moving parts of it out of wasi-libc seems like a good simplification, and I agree that having multiple duplicated `bh_` types seems unnecessary.

Reply:
I've added a separate header `platform_wasi.h` as suggested. I've also moved `wasi_errno.h` to shared platform sources and renamed it to `libc_errno.h` but there might be a better place/name for that. The additional changes (moving the WASI types around) are in a separate commit to make it a bit easier to review.

let me know if that looks good and hopefully this can be merged soon enough!